unused_variable suggets removing struct field from destructive binding which obv. does not work like this

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

Given the following code: <!-- Please provide a link to play.rust-lang.org -->

```rust
struct Struct {
    x: isize,
    y: isize
}

fn main() {
    match (Struct { x: 239, y: 240 }) {
        // ignored field
        Struct { x: unused, .. } => {}
    }
}
```

The current output is:

```
warning: unused variable: `unused`
 --> src/main.rs:9:21
  |
9 |         Struct { x: unused, .. } => {}
  |                     ^^^^^^-
  |                     |
  |                     help: try removing the field
  |
  = note: `#[warn(unused_variables)]` on by default

warning: field `y` is never read
 --> src/main.rs:3:5
  |
1 | struct Struct {
  |        ------ field in this struct
2 |     x: isize,
3 |     y: isize
  |     ^
  |
  = note: `#[warn(dead_code)]` on by default
```

The suggestion is questionable because  `Struct { x:  .. }`  is not valid rust code, we should probably suggest removing the `x: ` as well..?

Errors of the code rustc suggested:
````
The following errors were reported:
error: `..` patterns are not allowed here
 --> src/main.rs:9:22
  |
9 |         Struct { x:  .. } => {}
  |                      ^^
  |
  = note: only allowed in tuple, tuple struct, and slice patterns

error[E0027]: pattern does not mention field `y`
 --> src/main.rs:9:9
  |
9 |         Struct { x:  .. } => {}
  |         ^^^^^^^^^^^^^^^^^ missing field `y`
  |
help: include the missing field in the pattern
  |
9 |         Struct { x:  .., y } => {}
  |                        ~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
  |
9 |         Struct { x:  .., .. } => {}
  |                        ~~~~~~

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0027`.
Original diagnostics will follow.
````


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"chenyukang"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_passes/src/liveness.rs
+++ b/compiler/rustc_passes/src/liveness.rs
@@ -1548,7 +1548,13 @@ impl<'tcx> Liveness<'_, 'tcx> {
                 .or_insert_with(|| (ln, var, vec![id_and_sp]));
         });
 
-        let can_remove = matches!(&pat.kind, hir::PatKind::Struct(_, _, true));
+        let can_remove = match pat.kind {
+            hir::PatKind::Struct(_, fields, true) => {
+                // if all fields are shorthand, remove the struct field, otherwise, mark with _ as prefix
+                fields.iter().all(|f| f.is_shorthand)
+            }
+            _ => false,
+        };
 
         for (_, (ln, var, hir_ids_and_spans)) in vars {
             if self.used_on_entry(ln, var) {

--- a/src/test/ui/suggestions/try-removing-the-field.rs
+++ b/src/test/ui/suggestions/try-removing-the-field.rs
@@ -14,4 +14,19 @@ fn use_foo(x: Foo) -> i32 {
     return foo;
 }
 
+// issue #105028, suggest removing the field only for shorthand
+fn use_match(x: Foo) {
+    match x {
+        Foo { foo: unused, .. } => { //~ WARNING unused variable
+                                     //~| help: if this is intentional, prefix it with an underscore
+        }
+    }
+
+    match x {
+        Foo { foo, .. } => { //~ WARNING unused variable
+                             //~| help: try removing the field
+        }
+    }
+}
+
 fn main() {}

--- a/src/test/ui/suggestions/try-removing-the-field.stderr
+++ b/src/test/ui/suggestions/try-removing-the-field.stderr
@@ -8,5 +8,19 @@ LL |     let Foo { foo, bar, .. } = x;
    |
    = note: `#[warn(unused_variables)]` on by default
 
-warning: 1 warning emitted
+warning: unused variable: `unused`
+  --> $DIR/try-removing-the-field.rs:20:20
+   |
+LL |         Foo { foo: unused, .. } => {
+   |                    ^^^^^^ help: if this is intentional, prefix it with an underscore: `_unused`
+
+warning: unused variable: `foo`
+  --> $DIR/try-removing-the-field.rs:26:15
+   |
+LL |         Foo { foo, .. } => {
+   |               ^^^-
+   |               |
+   |               help: try removing the field
+
+warning: 3 warnings emitted
 

